File: V2/particles.js
function polygon(color, points) {
  // print("polygon", color, points[0], typeof points, typeof color);
  push();
  fill(...color);
  stroke(...color);
  // noStroke()
  // strokeWeight(10)
  beginShape();
  for (let point of points) {
    vertex(...point);
  }

  endShape();
  pop();
}

class Spark {
  //  By https://www.youtube.com/watch?v=wNMRq_uoWM0
  constructor(loc, angle, speed, color, scale = 1) {
    // console.log("this.color", this.color, color);

    this.loc = loc;
    this.angle = angle;
    this.speed = speed;
    this.scale = scale;
    this.color = color;
    this.alive = true;
  }

  point_towards(angle, rate) {
    let rotate_sign;
    rotate_direction = ((angle - this.angle + PI * 3) % (PI * 2)) - PI;
    try {
      rotate_sign = abs(rotate_direction) / rotate_direction;
    } catch {
      rotate_sing = 1;
    }
    if (abs(rotate_direction) < rate) {
      this.angle = angle;
    } else {
      this.angle += rate * rotate_sign;
    }
  }

  calculate_movement(dt) {
    return [
      cos(this.angle) * this.speed * dt,
      sin(this.angle) * this.speed * dt,
    ];
  }
  move(dt) {
    let movement = this.calculate_movement(dt);
    this.loc[0] += movement[0];
    this.loc[1] += movement[1];
    this.speed -= 0.085 + 0.025 * this.speed;
    if (this.speed <= 0) {
      this.alive = false;
    }
  }
  draw() {
    if (this.alive) {
      let points = [
        [
          this.loc[0] + cos(this.angle) * this.speed * this.scale,
          this.loc[1] + sin(this.angle) * this.speed * this.scale,
        ],
        [
          this.loc[0] +
            cos(this.angle + PI / 2) * this.speed * this.scale * 0.3,
          this.loc[1] +
            sin(this.angle + PI / 2) * this.speed * this.scale * 0.3,
        ],
        [
          this.loc[0] - cos(this.angle) * this.speed * this.scale * 3.5,
          this.loc[1] - sin(this.angle) * this.speed * this.scale * 3.5,
        ],
        [
          this.loc[0] +
            cos(this.angle - PI / 2) * this.speed * this.scale * 0.3,
          this.loc[1] -
            sin(this.angle + PI / 2) * this.speed * this.scale * 0.3,
        ],
      ];
      // console.log("this.color", this.color);
      polygon(this.color, points);
    }
  }
}

class Sparks {
  constructor() {
    this.particles = [];
  }
  update() {
    let l = this.particles.length;
    for (let i = 0; i < l; i++) {
      this.particles[l - i - 1].move(1);

      if (!this.particles[l - i - 1].alive) {
        this.particles.pop(l - i - 1);
      }
    }
  }
  create_particle(
    loc,
    num = 1,
    angle = null,
    speed = null,
    color = [254, 254, 254],
    scale = 2.1
  ) {
    //  num{ number of particles to create
    for (let i = 0; i < num; i++) {
      if (angle === null) {
        angle = radians(random(0, 360));
      }
      if (speed === null) {
        speed = random(3, 6);
      }
      let n = new Spark(loc, angle, speed, color, scale);
      this.particles.push(n);
    }
  }
    draw() {
      this.update()
    for (let spark of this.particles) {
      spark.draw();
    }
  }
}
let sparks 

function setup() {
  createCanvas(500, 500);
    frameRate(60);
    sparks = new Sparks();
}
function draw() {
  background(0);
    if (isPressed) {
        
        sparks.create_particle([mouseX, mouseY], num = 3);
        sparks.draw();
        
    }
//   for (var i = sparks.length - 1; i >= 0; i--) {
//     sparks[i].move(1);
//     sparks[i].draw();
//     if (!sparks[i].alive) {
//       sparks.pop(i);
//     }
//   }
}

// function keyPressed() {
//   print("keyPressed", mouseX, mouseY);
//   sparks.push(
//     new Spark(
//       [mouseX, mouseY],
//       radians(random(0, 360)),
//       random(3, 6),
//       [255, 255, 255],
//       2
//     )
//   );

//   sparks.push(
//     new Spark(
//       [mouseX, mouseY],
//       radians(random(0, 360)),
//       random(3, 6),
//       [255, 255, 255],
//       2
//     )
//   );

//   sparks.push(
//     new Spark(
//       [mouseX, mouseY],
//       radians(random(0, 360)),
//       random(3, 6),
//       [255, 255, 255],
//       2
//     )
//   );
// }
let isPressed = false;
function mousePressed(event) {
  if (event.button === 0) {
    isPressed = true;
  }
}
function mouseReleased(event) {
  if (event.button === 0) {
    isPressed = false;
  }
}

File: V1/src/particles.py
from pygame.locals import *
import pygame
import sys
import math
import random

class Spark():
    # By https://www.youtube.com/watch?v=wNMRq_uoWM0
    def __init__(self, loc, angle, speed, color, scale=1):
        self.loc = list(loc)
        self.angle = angle
        self.speed = speed
        self.scale = scale
        self.color = color
        self.alive = True

    def point_towards(self, angle, rate):
        rotate_direction = ((angle - self.angle + math.pi * 3) %
                            (math.pi * 2)) - math.pi
        try:
            rotate_sign = abs(rotate_direction) / rotate_direction
        except ZeroDivisionError:
            rotate_sing = 1
        if abs(rotate_direction) < rate:
            self.angle = angle
        else:
            self.angle += rate * rotate_sign

    def calculate_movement(self, dt):
        return [math.cos(self.angle) * self.speed * dt, math.sin(self.angle) * self.speed * dt]
        
    def move(self, dt):
        movement = self.calculate_movement(dt)
        self.loc[0] += movement[0]
        self.loc[1] += movement[1]
        self.speed -= 0.085+ 0.025*self.speed
        if self.speed <= 0:
            self.alive = False

    def draw(self, surf):
        if self.alive:
            points = [
                [self.loc[0] + math.cos(self.angle) * self.speed * self.scale,
                 self.loc[1] + math.sin(self.angle) * self.speed * self.scale],
                [self.loc[0] + math.cos(self.angle + math.pi / 2) * self.speed * self.scale * 0.3,
                 self.loc[1] + math.sin(self.angle + math.pi / 2) * self.speed * self.scale * 0.3],
                [self.loc[0] - math.cos(self.angle) * self.speed * self.scale * 3.5,
                 self.loc[1] - math.sin(self.angle) * self.speed * self.scale * 3.5],
                [self.loc[0] + math.cos(self.angle - math.pi / 2) * self.speed * self.scale * 0.3,
                 self.loc[1] - math.sin(self.angle + math.pi / 2) * self.speed * self.scale * 0.3],
            ]
            pygame.draw.polygon(surf, self.color, points)

class Sparks:
    def __init__(self):
        self.particles = []
    
    def update(self):
        l = len(self.particles)
        for i in range(l):
            self.particles[l-i-1].move(1)
            
            if not self.particles[l-i-1].alive:
                self.particles.pop(l-i-1)
                
    def create_particle(self, loc, num=1, angle=None, speed=None, color=(254, 254, 254), scale=2.1):
        # num: number of particles to create
        for i in range(num):
            if angle is None:
                angle = math.radians(random.randint(0, 360))
            if speed is None:
                speed = random.randint(3, 6)
            new = Spark(loc, angle, speed, color, scale)
            self.particles.append(new)

    def draw(self, surf):
        self.update()
        for spark in self.particles:
            spark.draw(surf)
 
if __name__ == '__main__':          
    sparks=[]           
    clock = pygame.time.Clock()
    pygame.init()
    screen = pygame.display.set_mode((500, 500))


    while True:
        pygame.display.update()
        screen.fill((0, 0, 0))
        clock.tick(60)
        
        for event in pygame.event.get():
            if event.type == QUIT:
                pygame.quit()
                sys.exit()
            if event.type == KEYDOWN:
                mx, my = pygame.mouse.get_pos()
                sparks.append(Spark([mx, my], math.radians(random.randint(
                    0, 360)), random.randint(3, 6), (255, 255, 255), 2))
                
                sparks.append(Spark([mx, my], math.radians(random.randint(
                    0, 360)), random.randint(3, 6), (255, 255, 255), 2))
                
                sparks.append(Spark([mx, my], math.radians(random.randint(
                    0, 360)), random.randint(3, 6), (255, 255, 255), 2))
                
        for i, spark in sorted(enumerate(sparks), reverse=True):
            spark.move(1)
            spark.draw(screen)
            if not spark.alive:
                sparks.pop(i)